### ISSUE TITLE
Do not introduce `result` in the codegen if it isnt needed. 

### ISSUE THREAD

**Issue by jmgomez** [nim-lang/Nim#22411]
### Summary

~Introduce a pragma that doesnt add a result to a function marked with it~

Do not introduce `result` in the codegen if it isnt needed. 

### Description

`result` is great but sometimes it is quite inconvenient. For instance, when you return an imported type it inits it like 
```
	TY__fPaIMQiBE59cMXjFvVSW8SQ result{};
```
Which can be problematic as it calls the default constructor and it may be deleted. 

Let me know if something like this would be accepted, I can work on it. 

### Alternatives

Attempt to initialise it from the list of known constructors

**Comment by juancarlospaco** [comment]
This is cpp only?, repro sample?.

**Comment by jmgomez** [comment]
> This is cpp only?, 

The functionality? No. Ideally it will allow for just not having a result but you can still return. 

> repro sample?

Repro what? The edge case where it will fail? It is just a type with a deleted ctor.

But yes, it is specially bad when interoping with cpp

**Comment by juancarlospaco** [comment]
> allow for just not having a result but you can still return.

`{.asmnostackframe.}` ?.

Also I think `{.noInit.}` on the proc declaration is supposed to omit the `result = ...`,
but in reality it does nothing, I dunno if its a bug (probably), maybe worth to investigate this,
I tried to use `{.noInit.}` but does nothing and then used `{.asmnostackframe.}` for JS interop.

In JS interop, sometimes you end up with codegen like (pseudocode), I dunno if that works for cpp:

```js
var result = null
return 42
return result
```

So basically you just "cut off" the flow with a `return value`.

**Comment by arnetheduck** [comment]
> Introduce a pragma 

I don't think a pragma is needed really - it could simply work in the way that if you don't touch `result` in your code, the codegen would just skip it - in fact, we write most of our code this way and we simply hope that the now pointless `result` variable will be optimized away - unfortunately, nim is [not yet](https://github.com/nim-lang/Nim/issues/19357) smart enough to do this but in theory it could work.

We've come to the point where for efficiency in hotspots we might write stuff like `proc f(): Type = result = \n <actual code>`, ie the use of `result` is just a piece of noise that we add to the function declaration line to work around codegen bugs but don't actually use it in the code (using expressions instead), but at the end of the day, it shouldn't have to be there - this would remove the need for `noinit` and other kludges as well.

**Comment by jmgomez** [comment]
> {.asmnostackframe.} ?.

Oh my! This seems to work, not sure yet as it breaks `virtual`, so gonna fix it and see. Wonder how many spells Im still missing :P  
I did try `noInit` and yes, it doenst do anything and not dont think it should do (as you will trigger the ctor just by making a declaration).

**Comment by jmgomez** [comment]
> I don't think a pragma is needed really - it could simply work in the way that if you don't touch result in your code, the codegen would just skip it

Yes, this would be much better.

**Comment by juancarlospaco** [comment]
> it breaks virtual

Then fix virtual + asmnostackframe and you are good to go... 🙂

If its really a blocker for you, I guess you can always workaround with `out` parameters.

**Comment by jmgomez** [comment]
Yes, I will fix `virtual` to play nice with `asmnostackframe` anyways. I mean, Im already doing a nasty workaround `codegenDecl` + `member` to wrap the C++ definition. `asmnostackframe` approach would be better but it's still hacky. And no, unfortunately `out` wouldnt work as this is an interfacing issue

What @arnetheduck said IMO is the way to go. If it gets approved I can attempt to do it

**Comment by jmgomez** [comment]
For the time being, a macro like this would kinda work:
```nim

macro noresult(prc:untyped) = 
  assert prc.kind == nnkProcDef
  prc.addPragma(ident "asmnostackframe")
  prc.body = 
    genAst(exp=prc.body): 
      let res {.inject.} = exp
      {.emit: "return res;".}
  prc

class FTestTabSpawner of FWorkflowTabFactory:
  proc createTabBody(info {. constcpp, byref .}: FWorkflowTabSpawnInfo): TSharedRef[SWidget] {.virtual, constcpp, override , noresult.} = 
    newUObject[UUserWidget]().takeWidget()
   
   ```
